java-checks/src/main/java/org/sonar/java/checks/NullShouldNotBeUsedWithOptionalCheck.java

```diff
@@ -0,0 +1,149 @@
+
+      // check that the method is not annotated with @Nullable
+      checkNullableOptionalAnnotation(method.modifiers(), "Methods with an \"Optional\" return type should not be \"@Nullable\".");
+
+      // check that the method does not return "null"
+      method.accept(new ReturnNullVisitor());
+    }
+
+    super.visitMethod(method);
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree binaryExpression) {
+    // check that an @Optional is not compared to "null"
+    if (binaryExpression.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
+      ExpressionTree left = binaryExpression.leftOperand();
+      ExpressionTree right = binaryExpression.rightOperand();
+      if ((isOptional(left) && isNull(right)) || (isNull(left) && isOptional(right))) {
+        context.reportIssue(this, binaryExpression, "Remove this null-check of an \"Optional\".");
+      }
+    }
+
+    super.visitBinaryExpression(binaryExpression);
+  }
+
+  @Override
+  public void visitVariable(VariableTree variable) {
+    if (variable.type().symbolType().is(OPTIONAL)) {
+      checkNullableOptionalAnnotation(variable.modifiers(), "\"Optional\" variables should not be \"@Nullable\".");
+    }
+
+    super.visitVariable(variable);
+  }
+
+  private class ReturnNullVisitor extends BaseTreeVisitor {
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree returnStatement) {
+      checkNull(returnStatement.expression());
+
+      super.visitReturnStatement(returnStatement);
+    }
+
+    @Override
+    public void visitConditionalExpression(ConditionalExpressionTree conditionalExpression) {
+      checkNull(conditionalExpression.trueExpression());
+      checkNull(conditionalExpression.falseExpression());
+
+      super.visitConditionalExpression(conditionalExpression);
+    }
+
+    private boolean checkNull(ExpressionTree expression) {
+      if (isNull(expression)) {
+        context.reportIssue(NullShouldNotBeUsedWithOptionalCheck.this, expression, "Methods with an \"Optional\" return type should never return null.");
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public void visitLambdaExpression(LambdaExpressionTree lambdaExpressionTree) {
+      // don't visit lambdas, as they are allowed to return null
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      // don't visit inner class tree, as methods in there will be visited by outer class
+    }
+
+  }
+
+  private static boolean returnsOptional(MethodTree method) {
+    return method.returnType().symbolType().is(OPTIONAL);
+  }
+
+  private static boolean isOptional(ExpressionTree expression) {
+    return expression.symbolType().is(OPTIONAL) && !isNull(expression);
+  }
+
+  private static boolean isNull(ExpressionTree expression) {
+    return expression.is(Tree.Kind.NULL_LITERAL);
+  }
+
+  private void checkNullableOptionalAnnotation(ModifiersTree modifiers, String message) {
```

Comment:
This method should be called `checkNullableAnnotation`, as you already know that it is applied on a `Nullable` method or variable.